### PR TITLE
port `square` to structured

### PR DESCRIPTION
**Pull request by eqy**
#55070

`square` really just calls `pow` so it seems a separate `TensorIterator` doesn't need to be built for it

### CODE REVIEW

**Comment by eqy** [comment]
Does porting to structure require a derivative to be registered for square when previously it wasn't in `derivatives.yaml`?

File: aten/src/ATen/native/UnaryOps.cpp [hunk 1]
```diff
@@ -433,9 +438,9 @@ Tensor& arctanh_out(const Tensor& self, Tensor& result) { return at::atanh_out(r
 Tensor arctanh(const Tensor& self) { return self.atanh(); }
 Tensor& arctanh_(Tensor& self) { return self.atanh_(); }
 
-Tensor& square_out(const Tensor& self, Tensor& result) { return at::pow_out(result, self, 2); }
-Tensor square(const Tensor& self) { return at::pow(self, 2); }
-Tensor& square_(Tensor& self) { return self.pow_(2); }
+// FIXME: remove const_cast once pow_out is ported to structured
```
Comment (bdhirsh) [inline aten/src/ATen/native/UnaryOps.cpp:441]: Actually, `pow` is already structured. The const_cast is needed because the C++ API signatures generated in `build/aten/src/ATen/Functions.h` all return mutable `Tensor&` instead of `const Tensor&`. We have an open issue for it here: https://github.com/pytorch/pytorch/issues/54001

**Comment by bdhirsh** [comment]
> Looks like I broke derivatives :/

Oh, this one's interesting. `square` LOOKS like it's a good structured kernel candidate, since it has all 3 variants and a CPU/CUDA dispatch entry for the `out` variant. But it looks like that entry is wrong; `square_out` is actually a mathematically composite kernel, since its implementation just calls into `at::pow_out`. That's why it was able to get away without having any derivatives.yaml entries. By trying to make it structured, the codegen will try to register cpu/cuda-specific implementations of `square` and `square_`, which assume that there's an actual autograd formula.

I think that you can just delete the dispatch table entry of `square_out` in native_functions.yaml, which will tell the codegen that all 3 ops are actually composite kernels:
```
  dispatch:
    CPU, CUDA: square_out
```

**Comment by eqy** [comment]
Thanks! I was confused as to how moving to structured could have broke derivatives if it didn't need one registered previously. Interestingly adding a `square` entry in `derivatives.yaml` didn't seem to fix it but I'm not familiar with the intricacies of registering a function there.

**Comment by eqy** [comment]
It looks like without the dispatch table codegen doesn't actually generate a function definition. This seems tricky as I can't find any examples of structured functions without dispatch table either.

**Comment by bdhirsh** [comment]
> It looks like without the dispatch table codegen doesn't actually generate a function definition. This seems tricky as I can't find any examples of structured functions without dispatch table either.

Oh, you can fix that by manually specifying that the operator is composite. So you can update the dispatch table to:
```
 dispatch:
    CompositeImplicitAutograd: square_out
```

Right now the codegen assumes that your op is composite if you don't specify a dispatch entry, but it looks like it only does that for unstructured operators ([here](https://github.com/pytorch/pytorch/blob/0d11dbf5119e1e4c69016ed1527bcf5013b208af/tools/codegen/model.py#L351)). @ezyang is there any reason we don't allow defaulting the dispatch table for structured ops? I know we have some logic to auto-generate composite kernels for structured ops in some cases, but I don't think that would interfere with the defaulting here.

**Comment by eqy** [comment]
codegen doesn't seem to like that on my end:
`AssertionError: Do not explicitly specify CompositeExplicitAutograd dispatch key on structured functions, they will be automatically generated for you`

**Comment by bdhirsh** [comment]
> codegen doesn't seem to like that on my end:
> `AssertionError: Do not explicitly specify CompositeExplicitAutograd dispatch key on structured functions, they will be automatically generated for you`

Ah, you want `CompositeImplicitAutograd`, not `CompositeExplicitAutograd`. the Implicit is the part that means the op is pasing off the work of dealing with autograd to whatever ops it calls, which is what we want to do here (`square` should let `pow` figure out how to properly deal with the backward pass).

Yeah, the codegen does some fancy autogeneration stuff for `CompositeExplicitAutograd` in some cases, which is why it throws you an error when you write it yourself for structured ops.

**Comment by bdhirsh** [comment]
@eqy The fix is a one line change (see [here](https://github.com/pytorch/pytorch/pull/58336/files#diff-a34ee043aa7bde7b608aa03bdd3ff6d24af2fea53f96c01be304034c30a8bfcdL351))- do you want to add it in to your existing PR? That'll test the fix more fully.

**Comment by eqy** [comment]
> @eqy The fix is a one line change (see [here](https://github.com/pytorch/pytorch/pull/58336/files#diff-a34ee043aa7bde7b608aa03bdd3ff6d24af2fea53f96c01be304034c30a8bfcdL351))- do you want to add it in to your existing PR? That'll test the fix more fully.

On my end that seems to fire an assert:
```
    assert dispatch != {DispatchKey.CompositeImplicitAutograd: cpp.name(func)}, \                                     
    AssertionError: unnecessary dispatch table for this function; just delete the dispatch key entirely                   
      in /nvdl/pytorch/cmake/../aten/src/ATen/native/native_functions.yaml:3893:                                          
          square.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)                                                          
```

**Comment by bdhirsh** [comment]
> On my end that seems to fire an assert:
> 
> ```
>     assert dispatch != {DispatchKey.CompositeImplicitAutograd: cpp.name(func)}, \                                     
>     AssertionError: unnecessary dispatch table for this function; just delete the dispatch key entirely                   
>       in /nvdl/pytorch/cmake/../aten/src/ATen/native/native_functions.yaml:3893:                                          
>           square.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)                                                          
> ```


That fix will allow you to not have specify a dispatch table entry if your op is mathematically composite. So you should just be able to remove the line entirely:
```
 dispatch:
    CompositeImplicitAutograd: square_out
```

**Comment by eqy** [comment]
> > On my end that seems to fire an assert:
> > ```
> >     assert dispatch != {DispatchKey.CompositeImplicitAutograd: cpp.name(func)}, \                                     
> >     AssertionError: unnecessary dispatch table for this function; just delete the dispatch key entirely                   
> >       in /nvdl/pytorch/cmake/../aten/src/ATen/native/native_functions.yaml:3893:                                          
> >           square.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)                                                          
> > ```
> 
> That fix will allow you to not have specify a dispatch table entry if your op is mathematically composite. So you should just be able to remove the line entirely, and still get a generated function definition:
> 
> ```
>  dispatch:
>     CompositeImplicitAutograd: square_out
> ```

Interesting, I now get
```
  File "/nvdl/pytorch/tools/codegen/model.py", line 485, in __post_init__
    assert not self.out.has_composite_implicit_autograd_kernel
    AssertionError
```
without a dispatch table.

**Comment by bdhirsh** [comment]
> Interesting, I now get
> 
> ```
>   File "[...]pytorch/tools/codegen/model.py", line 485, in __post_init__
>     assert not self.out.has_composite_implicit_autograd_kernel
>     AssertionError
> ```
> 
> without a dispatch table.


Ah sorry about that, now I see from [the comment in the codegen](https://github.com/pytorch/pytorch/blob/master/tools/codegen/model.py#L483) that there isn't actually support for mathematically composite structured kernels yet. I'm going to play around with your code locally and see what we need to get it to work- we already have a similar version working (the `CompositeExplicitAutograd` key), so hopefully there won't be too much work involved.

**Comment by ezyang** [comment]
@bdhirsh OK, this is why I asked you to exercise the case ;)

I don't think that making codegen work here is the right call. The problem is the implementation of square here:

```
// FIXME: remove const_cast once pow_out's generated signature is updated
TORCH_IMPL_FUNC(square_out)
(const Tensor& self, const Tensor& result) { at::pow_out(const_cast<Tensor&>(result), self, 2); }
```

Your implementation function calls to `pow_out` (dispatched!), which doesn't support autograd; none of our out variants support structured. So in codegen, to make `square` work, you'd have to gen `at::pow(self, 2)` but there's no way for codegen to know that this is what is codegened.

In other words, you fell afoul of https://github.com/pytorch/pytorch/issues/50953 . Without that issue being fixed, the workaround here is just not make square structured and keep it the way it was before (sorry @eqy).

cc @mruberry @ngimel

**Comment by bdhirsh** [comment]
> Your implementation function calls to `pow_out` (dispatched!), which doesn't support autograd; none of our out variants support structured. So in codegen, to make `square` work, you'd have to gen `at::pow(self, 2)` but there's no way for codegen to know that this is what is codegened.
> 
> In other words, you fell afoul of #50953 . Without that issue being fixed, the workaround here is just not make square structured and keep it the way it was before

Synced about this in a 1:1, that makes sense to me now. Originally, I thought that #50953 only applied to overloads of an operator, not all composite ops. I thought that it would be fine to use structured kernels to go `at::square()` -> `square_out()` -> `pow_out()`, but that doesn't work because `pow_out()`, and `out` ops in general, don't support autograd (and `at::square()` expects it to; since it's mathematically composite, it delegates autograd responsibility to its base ops.
